### PR TITLE
Adding extra hooks to be triggered on external modules management operations

### PR DESCRIPTION
**Pull request by tbembersimeao**
Implements hooks described in the following issue: https://github.com/vanderbilt/redcap-external-modules/issues/13#issuecomment-323105463.

### CODE REVIEW

**Comment by tbembersimeao** [comment]
@mmcev106 do you have plans to to review this PR soon? I totally understand that your team might be busy with priority issues and I don't want to rush you, I just wanted to know an estimate since my team needs this feature to develop an external module.

**Comment by taylorr4** [comment]
@tbembersimeao, given the nature of your PR, i think this one may take @mmcev106 and the Data Core Team a bit more time to consider. What you're proposing is not just a simple code change or modification of functionality, but you're introducing completely new concepts that didn't exist before (i.e., External Modules having their own hooks). While REDCap has hooks, i don't believe i've heard of the idea of EM having its own version of hooks. It's interesting. So while i'm sure these additions will be useful to you and your team, lots of things have to be weighed (like with REDCap itself) to consider whether this all makes sense with regard to the current development path of EM, and whether or not it will be useful to most module developers. 

So in summary, i ask for some patience on this one as i'm sure some things will need to be discussed on the development side (most likely after the holidays) for @mmcev106 and other developers on that team to discern whether this kind of functionality makes sense in the whole grand scheme of things for EM. Thanks!

**Comment by tbembersimeao** [comment]
@taylorr4 thank you for your quick response. Our time frame is flexible, so np at all.

I understand that this not a simple request and it brings questions to be considered, but I just want to advocate one more time that it is important - although not urgent - to provide to developers, either by the proposed approach or not, a way to act on module management events such as enable and disable.

Let's take the use case of a module that needs a new db table to store its own stuff, which is a widespread case in modular web systems like Wordpress and Drupal. Without this feature, the only way to make this happen is to instruct users (sometimes non-technical ones) to run a SQL script manually before enabling the module - which makes the module less modular, less user friendly and more error prone.

FYI my use case is an API client module that needs to log API requests into a new db table.

Again, I do not consider that a top priority for the moment. But as the community grows and new external modules pop up, I believe this can be really useful.

**Comment by mmcev106** [comment]
I totally agree on the importance of this feature.  I will at least give you a status report within the next couple of weeks.

**Comment by 123andy** [comment]
Hey @tbembersimeao - re api logging - I have a custom handler that you might find useful:

We use a REDCap project with rules that allow/reject API access based on network, project, or user.  This allows us to open up our VPN networks to everyone but limit external access based on approved projects as a means to mitigate risk a little bit...  You could argue this doesn't do much, but FWIW...

Here is my custom /api/index.php
```
<?php
/*****************************************************************************************
 **  REDCap is only available through ACADMEMIC USER LICENSE with Vanderbilt University
 ******************************************************************************************/

// Set error reporting and prevent caching
error_reporting(E_ALL); //0);

// Prevent caching
header("Expires: 0");
header("cache-control: no-store, no-cache, must-revalidate");
header("Pragma: no-cache");

//Connect to db
include dirname(dirname(__FILE__)) . "/database.php";

$db = new mysqli($hostname, $username, $password, $db);
if ($db->connect_errno) {
    exit("The hostname ($hostname) / username ($username) / password (XXXXXX) combination could not connect to the MySQL server. 
		Please check their values.");
}

/*
 * Taking access control functionality out of apache config file and including it here.
 * For now, if there is no token passed request goes through.
 * If a token is passed, validate request using whitelist rules defined in redcap
 * All successfull and failed requests are logged in the redcap_api_custom_log table
*/

require_once('whitelist.php');
$wl = new whitelist($db);

// IF PASS - EXECUTE REQUEST
if ($wl->evaluate()) {
    
    $file = dirname(dirname(__FILE__)) . "/redcap_v" . $wl->redcap_version . "/API/index.php";

    if (file_exists($file)) {
        $wl->log();
        include $file;
    } else {
        $wl->result = "ERROR";
        $wl->comment = "No REDCap directories found at $file. Please install the REDCap software and try again.";
        $wl->log();
        exit("No REDCap directories found. Please install the REDCap software and try again.");
    }

    // LOG THE COMPLETION OF THE API REQUEST (FOR ACCURATE DURATION)
    // Currently there is a problem that many of the API methods have an 'exit' in them and do not make it back from the include
    // command...  I will try to get these changed so they do return as it allows for much better logging
    $wl->logDuration();
}
```

Here is the whitelist class:
```
<?php

/**
* A php-based whitelist that permits access to the API based on IP and user/project rules
* 
* See https://stanfordmedicine.box.com/s/hjz2ohr4gc47xdzcv6hcn6hp3zpscj0a for the redcap data dictionary
* 
* SYNTAX TO CREATE LOG TABLE

CREATE TABLE `redcap_log_api_custom` (
  `log_id` int(10) NOT NULL AUTO_INCREMENT,
  `ip_address` varchar(50) COLLATE utf8_unicode_ci DEFAULT NULL,
  `username` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `project_id` int(10) DEFAULT NULL,
  `ts` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
  `duration` float DEFAULT NULL,
  `result` enum('PASS','REJECT','ERROR') CHARACTER SET utf8 DEFAULT NULL,
  `rule_id` int(10) DEFAULT NULL,
  `comment` text CHARACTER SET utf8,
  PRIMARY KEY (`log_id`)
) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;


// Once you have created the REDCap that stores the Whitelist, you can insert it into the config table:

INSERT INTO redcap_config (field_name, value) VALUES ('redcap_log_api_whitelist_pid', XXX);


*/
class whitelist
{
    private $rule_project_id;   // This is the redcap project that contains the rule data dictionary
    private $rules = array();   // Rules pulled from the REDCap Whitelist Project
    public $redcap_version;     // This is the redcap version number

    private $ts_start;          // This is the starting microtime
    private $db;                // This carries the mysqli db

    private $token;             // The API Token
    private $username;          // The TOKEN username
    private $project_id;        // The TOKEN project_id

    private $ip;                // This is the $_REQUEST ip address
    private $rule_id;           // The matching rule_id for logging

    public $log_id;
    public $result = "ERROR";   // PASS, REJECT, or ERROR
    public $comment;            // An optional annotation to be added to the log

    public $rejection_message = <<<EOT
ERROR: Stanford REDCap API Access Restricted

API Access (including the REDCap Mobile App) from untrusted networks requires:

1) an active VPN connection (see https://vpn.stanford.edu for VPN instructions)
    OR
2) an approved project/user-based API Exemption.

To connect now, confiure a VPN connection on your device or if there is a reason this is not possible,
contact us for more information at redcap-help@lists.stanford.edu or review https://medwiki.stanford.edu/x/hAtrAg
EOT;

    public function __construct(mysqli $db)
    {
        $this->ts_start = microtime(true);
        $this->ip = trim($_SERVER['REMOTE_ADDR']);
        $this->db = $db;
        
        $this->getRuleProjectId();
        $this->getRedcapVersion();
        $this->loadRules();
        $this->loadToken();
    }


    public function getRedcapVersion() {
        $sql = "SELECT value FROM redcap_config WHERE field_name = 'redcap_version'";
        if (!$result = $this->db->query($sql)) { die("Error running query at " . __LINE__ . " of " . __FILE__); }
        if ($result->num_rows == 0) { die("ERROR: redcap_version is not configured"); }
        $row = $result->fetch_assoc();
        $this->redcap_version = $row['value'];
    }

    
    public function getRuleProjectId() {
        $sql = "SELECT value FROM redcap_config WHERE field_name = 'redcap_log_api_whitelist_pid'";
        if (!$result = $this->db->query($sql)) { die("Error running query at " . __LINE__ . " of " . __FILE__); }
        if ($result->num_rows == 0) { die("ERROR: redcap_log_api_whitelist_pid is not configured"); }
        $row = $result->fetch_assoc();
        $this->rule_project_id = $row['value'];
    }
    

    public function evaluate() {
        // Step 1: Is the token missing
        if (empty($this->token)) {
            $this->comment  = "[No Token]: " . json_encode($_REQUEST);
            $this->result   = "PASS";
            return true;
        }

        // Load the project and username from the token
        $this->loadProjectUsername();
          
        // Step 2: Is the IP Whitelisted
        if (self::validIpAddress()) {
            $this->result    = "PASS";
            return true;
        }

        // Step 3: Is the username/project_id whitelisted
        if (self::validProjectUsername()) {
            $this->result    = "PASS";
            return true;
        }

        // Step 4: Failed all rules - reject
        $this->result        = "REJECT";
        self::log();

        header('HTTP/1.0 401 Unauthorized');
        exit($this->rejection_message . PHP_EOL . "If you are connected to the VPN and still receiving this message, please " .
            "contact us and include your current IP (" . $this->ip . ").");
    }


    public function validIpAddress() {
        foreach ($this->rules as $rule) {
            if (!empty($rule['ip_address'])) {
                if (self::ipCIDRCheck($rule['ip_address'])) {
                    $this->rule_id = $rule['rule_id'];
                    return true;
                }
            }
        }
        return false;
    }


    public function validProjectUsername() {
        // Skip if we don't have a valid username and project_id
        if (empty($this->username) OR empty($this->project_id)) return false;
        
        // Go through each rule and check for a match
        foreach ($this->rules as $rule) {
            if (empty($rule['username']) AND empty($rule['project_id'])) continue;
            
            if ($rule['username'] === $this->username AND $rule['project_id'] == $this->project_id) {
                // Allow based on username and project_id match
                $this->rule_id = $rule['rule_id'];
                return true;
            }
            
            if ($rule['project_id'] == $this->project_id AND empty($rule['username'])) {
                // Allow based on project_id
                $this->rule_id = $rule['rule_id'];
                return true;
            }
            
            if ($rule['username'] === $this->username AND empty($rule['project_id'])) {
                // Allow based on username role
                $this->rule_id = $rule['rule_id'];
                return true;
            }
        }
        return false;
    }


    public function loadRules() {
        $sql = "
            SELECT * 
            FROM (
                SELECT
                    a.record as rule_id,
                    max(if(a.field_name = 'username', a.value, NULL)) as username,
                    max(if(a.field_name = 'project_id', a.value, NULL)) as project_id,
                    max(if(a.field_name = 'ip_address', a.value, NULL)) as ip_address,
                    max(if(a.field_name = 'inactive', a.value, 0)) as inactive
                FROM
                    redcap_data a
                WHERE
                    a.project_id = " . intval($this->rule_project_id) . " 
                    AND a.record <> ''
                GROUP BY
                    a.record
            ) b
            WHERE
            b.inactive != 1;
        ";
        if (!$result = $this->db->query($sql)) { die("Error running query at " . __LINE__ . " of " . __FILE__); }
        
        while ($row = $result->fetch_assoc()) {
            $this->rules[] = $row;
        }
    }


    public function loadToken() {
        $this->token = isset($_REQUEST['token']) ? trim($_REQUEST['token']) : "";
    }


    public function loadProjectUsername() {
        $sql = "
            SELECT username, project_id 
            FROM redcap_user_rights
            WHERE api_token = ?";
        $stmt = $this->db->prepare($sql);
        $token = trim($this->token);
        $stmt->bind_param('s', $token);
        $stmt->execute();
        $stmt->store_result();
        if($stmt->num_rows > 0) {
            $stmt->bind_result($username, $project_id);
            while ($row = $stmt->fetch()) {
                $this->username = $username;
                $this->project_id = $project_id;
            }           
        } else {
            $this->comment  = "[Invalid Token]: " . json_encode($_REQUEST);
        }
        $stmt->free_result();
        $stmt->close();
    }


    // Checks if the IP is valid given an IP or CIDR range
    // e.g. 192.168.123.1 = 192.168.123.1/30
    public function ipCIDRCheck ($CIDR) {
        $ip = $this->ip;
        if(strpos($CIDR, "/") === false) $CIDR .= "/32";
        list ($net, $mask) = explode("/", $CIDR);
        $ip_net  = ip2long($net);
        $ip_mask = ~((1 << (32 - $mask)) - 1);
        $ip_ip = ip2long($ip);
        $ip_ip_net = $ip_ip & $ip_mask;
        return ($ip_ip_net == $ip_net);
    }

    // Since not all API commands are returning to the calling script, we write the log and update the duration later if it returns
    public function logDuration() {
        $duration = microtime(true) - $this->ts_start;
        $sql = "UPDATE redcap_log_api_custom SET duration = ? WHERE log_id = ?";
        $stmt = $this->db->prepare($sql);
        $stmt->bind_param('di', $duration, $this->log_id);
        $stmt->execute();
        $stmt->close();
    }


    public function log()
    {
        //$comments = empty($this->comment) ? implode(",",array_keys($_REQUEST)) : $this->comment;
        // For now, let's do thorough logging, but we can back off after a while if need be.
        $comments = empty($this->comment) ? json_encode($_REQUEST) : $this->comment;
        $sql = "INSERT INTO redcap_log_api_custom SET 
            ip_address = ?,
            username = ?,
            project_id = ?,
            result = ?,
            rule_id = ?,
            comment = ?";
        $stmt = $this->db->prepare($sql);
        $stmt->bind_param("ssisis",
            $this->ip,
            $this->username,
            $this->project_id,
            $this->result,
            $this->rule_id,
            $comments
        );
        $stmt->execute();
        
        // Record the log ID so we can update the duration later (if it returns)
        $this->log_id = $this->db->insert_id;
        $stmt->close();
    }
    
    public function debug() {
        print "<pre>" . print_r($this,true)."</pre>";
    }
}
```

**Comment by tbembersimeao** [comment]
Thank you @mmcev106!

And thank you @123andy for your code. My API integration module is quite done already, but anyways I really appreciate it - it can be useful in the future, since we are in the same scope of security concerns.

File: classes/ExternalModules.php [hunk 1]
```diff
@@ -1455,7 +1475,7 @@ private static function getEnabledModuleVersionsForProject($pid)
 
 	private static function shouldExcludeModule($prefix, $version)
 	{
-		if(strpos($_SERVER['REQUEST_URI'], '/manager/ajax/enable-module.php') !== false && $prefix == $_POST['prefix']){
+		if (strpos($_SERVER['REQUEST_URI'], '/manager/ajax/enable-module.php') !== false && $prefix == $_POST['prefix'] && $_POST['version'] != $version) {
```
Comment (mmcev106) [inline classes/ExternalModules.php]: Do you happen to remember why this change was necessary?

File: classes/ExternalModules.php [hunk 2]
```diff
@@ -1497,11 +1517,10 @@ private static function cacheAllEnableData()
 				$key = $row['key'];
 				$value = $row['value'];
 
-				if(self::shouldExcludeModule($prefix, self::KEY_VERSION)){
-					continue;
-				}
-
 				if($key == self::KEY_VERSION){
+					if (self::shouldExcludeModule($prefix, $value)) {
```
Comment (mmcev106) [inline classes/ExternalModules.php]: ...and this.  It seems like this would populate inconsistent enable data, but I suspect I may not have fully grasped the reasoning here yet.

File: docs/official-documentation.md [hunk 3]
```diff
@@ -143,10 +143,51 @@ class HideHomePageEmails extends AbstractExternalModule
...
+- `redcap_module_project_enable($prefix, $version, $project_id)`: triggered when a module gets enabled on a specific project
+- `redcap_module_project_disable($prefix, $version, $project_id)`: triggered when a module gets disabled on a specific project
+
+Examples:
+
+``` php
+<?php
+
+function redcap_module_system_enable($prefix, $version) {
+    if ($this->PREFIX != $prefix) {
```
Comment (mmcev106) [inline docs/official-documentation.md]: I think I might be missing something on the goal of checking the prefix here.  Shouldn't $this->PREFIX always match $prefix?  Actually, do we need to pass the prefix to these new hooks at all?

**Comment by mmcev106** [comment]
I finally got around to reviewing this PR, and it looks fantastic!  Thanks very much for submitting it!  I added a couple of comments/questions to specific lines, but I don't expect them to lead to any significant changes (if any).

File: classes/ExternalModules.php (hunk 1 above)
Comment (tbembersimeao) [reply classes/ExternalModules.php]: I am relaxing the exclude criteria because I need the module to be loaded at this initial moment - otherwise the enable (or module change) event will be missed, and the hooks won't be called.

This extra condition makes sure that only the old version is excluded, so we are still avoiding possible conflicts (see comments inside the if clause).

File: classes/ExternalModules.php (hunk 2 above)
Comment (tbembersimeao) [reply classes/ExternalModules.php]: `shouldExcludeModule()` expects a version _value_. Previously, the version _key_ was being passed. I fixed that because I needed this parameter to relax a condition (see my answer to the first comment) in order to make sure the enabled module is able to react to its own birth.

File: docs/official-documentation.md (hunk 3 above)
Comment (tbembersimeao) [reply docs/official-documentation.md]: Maybe this is a bad design, but all modules are able to react to this event, not only the enabled one. I remember I did that for convenience, in order to make use of `callHook()` functionality without changing it.

If this approach is not good enough, we could change `callHook()` or `startHook()` in order to prevent other modules to execute these hooks when the event is none of their business.

Comment (tbembersimeao) [reply docs/official-documentation.md]: I am working on a fix for that, I'll update the PR soon.

File: classes/ExternalModules.php (hunk 2 above)
Comment (mmcev106) [reply classes/ExternalModules.php]: I think my main concern here is that moving the exclude call inside the version check means that $projectEnableOverrides and $projectEnabledDefaults will now get populated for excluded modules (where they didn't previously).  Off the top of my head, I don't know if this will result in any actual problems.  It just seems like it makes the cached data logically inconsistent, which seems like something we should avoid if possible.

File: docs/official-documentation.md (hunk 3 above)
Comment (mmcev106) [reply docs/official-documentation.md]: Thanks!  I really like the idea of moving this check to `callHook()` or `startHook()` to abstract this away from module authors.

File: classes/ExternalModules.php (hunk 2 above)
Comment (tbembersimeao) [reply classes/ExternalModules.php]: You are right. I've just committed a fix for that.

File: docs/official-documentation.md (hunk 3 above)
Comment (tbembersimeao) [reply docs/official-documentation.md]: Done.

**Comment by tbembersimeao** [comment]
@mmcev106 thanks for the feedback! I've made a few fixes based on your comments.

File: classes/ExternalModules.php (hunk 1 above)
Comment (mmcev106) [reply classes/ExternalModules.php]: As is, I believe this will cause the enable request to crash when upgrading modules with composer dependencies if the ComposerAutoloaderInit class hasn't changed between module versions.  I suspect we may need a different solution to get around this problem (perhaps a separate ajax request).  It may take me a while to get around to it, but I'd be happy to find and implement a fix for this.

Comment (mmcev106) [reply classes/ExternalModules.php]: On second thought, it doesn't seem like this class loading issue would be limited to that composer class...  I might have to do more testing around this case.

Comment (tbembersimeao) [reply classes/ExternalModules.php]: How can I reproduce this error? I tried to shift versions of a module containing (and including) the same `vendor/autoload.php` file and I got no error, even for the test in which I return `false` all the times on `shouldExcludeModule()`. Am I missing something?

**Comment by tbembersimeao** [comment]
@mmcev106, I asked you a question in one of our outdated code diff threads that might has been missed. Could you please give me a direction on this? Thanks!

**Comment by mmcev106** [comment]
Was it on ComposerAutoloaderInit thread?  If so, I don't have an answer to that one yet.  It's still on my TODO list, but I'm going to have to find some time to investigate it first before I'll have more info.

**Comment by tbembersimeao** [comment]
@mmcev106 yes, I was just making sure that it wasn't lost among the other notifications/threads/issues/PRs. Thanks!

File: classes/ExternalModules.php (hunk 1 above)
Comment (mmcev106) [reply classes/ExternalModules.php]: Please disregard my comments here, I believe this change works just fine.  Module loading has changed since that comment was added.  The ComposerAutoloaderInit scenario is no longer an issue.  I've made a note to update that comment after this PR is merged.

**Comment by mmcev106** [comment]
Thanks for the hard work!  This is an awesome addition.
